include/swift/Basic/TypeID.h

```diff
@@ -0,0 +1,99 @@
+//===--- TypeID.h - Simple Type Identification ------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file defines the TypeID template, which provides a numeric
+//  encoding of (static) type information for use as a simple replacement
+//  for run-time type information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_BASIC_TYPEID_H
+#define SWIFT_BASIC_TYPEID_H
+
+#include <cstdint>
+#include <vector>
+
+namespace swift {
+
+/// Form a unique 64-bit integer value describing the type `T`.
+///
+/// This template needs to be specialized for every type that can
+/// participate in this kind of run-time type information, e.g., so
+/// that it can be stored in \c AnyValue.
+template<typename T>
+struct TypeID;
+
+/// Each type "zone" can contain 256 separate types for use in typeid,
+/// which should be enumerated.
+struct TypeIdZones {
+  enum ZoneValues : uint8_t {
+    /// C/C++ language and standard library types.
+    C = 0,
+
+    /// A zone used only for unit testing.
+    UnitTests = 1,
+
+    /// A zone used for the Swift type checker.
+    TypeChecker,
+  };
+};
+
+/// Form a type ID given a zone and type value.
+constexpr uint8_t formTypeID(uint8_t zone, uint8_t type) {
+  return (zone << 8) | type;
+}
+
+/// Assign the given type a particular value in the zone.
+#define SWIFT_TYPEID(Zone, Type, Value)                           \
+template<> struct TypeID<Type> {                                        \
+  static const uint64_t value = formTypeID(TypeIdZones::Zone, Value);   \
+  static const unsigned depth = 0;                                      \
+}
+
+/// Assign the given single-param template a particular value in the zone.
+#define SWIFT_TYPEID_TEMPLATE1(Zone, Template, Value, Param1, Arg1)     \
+template<Param1> struct TypeID<Template<Arg1>> {                        \
+  static const unsigned depth = TypeID<Arg1>::depth + 1;                \
+  static_assert(depth < 4, "cannot fit type in 64-bit value");          \
+                                                                        \
+private:                                                                \
+ static const uint64_t templateID = formTypeID(TypeIdZones::Zone, Value); \
+                                                                        \
+public:                                                                 \
+  static const uint64_t value =                                         \
+    (templateID << (depth * 8)) | TypeID<Arg1>::value;                  \
```

Comment:
Shouldn't this be `depth * 16`. Also, could we could possibly avoid `depth` all together using `(TypeID<Arg1>::value << 16) | templateID`, assuming `value` is guaranteed to be non-zero (and/or `static_assert` that it's non-zero)?

Reply:
Good idea, I've done this.